File: Daily Exercises/Mar 2021/2021-03-28 : 173. Binary Search Tree Iterator/173. Binary Search Tree Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.s=[root]
        while(self.s[-1].left):
            self.s.append(self.s[-1].left)


    def next(self) -> int:        
        ans=self.s.pop()
        if ans.right:
            self.s.append(ans.right)
            while(self.s[-1].left):
                self.s.append(self.s[-1].left)
        return ans.val
        

    def hasNext(self) -> bool:
        return bool(self.s)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Daily Exercises/Mar 2021/2021-03-28 : 173. Binary Search Tree Iterator/173. Binary Search Tree Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
public:
    stack<TreeNode*> s;
    BSTIterator(TreeNode* root) {
        s.push(root);
        while(s.top()->left!=nullptr){
            s.push(s.top()->left);
        }
    }
    
    int next() {
        TreeNode* ans=s.top();
        s.pop();
        if(ans->right!=nullptr){
            s.push(ans->right);
            while(s.top()->left!=nullptr){
            s.push(s.top()->left);
            }
        }
        return ans->val;
    }
    
    bool hasNext() {
        return !s.empty();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */
